File: TopologicalSort/802. Find Eventual Safe States/802. Find Eventual Safe States-DFS.cpp
class Solution {
    // bfs to find out degree
    vector<int>visited;
    vector<vector<int>>graph;
public:
    vector<int> eventualSafeNodes(vector<vector<int>>& graph) {
        int n = graph.size();
        visited.resize(n);
        this->graph = graph;
        
        vector<int>ret;
        for (int i = 0; i < n; i++) {
            if ( DFS(i) == true ) ret.push_back(i);
        }
        return ret;
    }
    
    bool DFS(int cur) {
        if (visited[cur] == 1) return true;
        if (visited[cur] == 2) return false; // has loop
        
        visited[cur] = 2;
        for (int next : graph[cur]) 
            if ( DFS(next) == false ) return false;
        
        visited[cur] = 1; // ensure no loop
        return true;
    }
};

// DFS: visited[i]
// 0: never visited;
// 2: first-time visit unsafe status;
// 1: permanently visited, safe.

File: TopologicalSort/802. Find Eventual Safe States/802. Find Eventual Safe States-DFS.java
class Solution {
    // topological sort
    List<Integer> rets = new ArrayList<>();
    public List<Integer> eventualSafeNodes(int[][] graph) {
        int n = graph.length;
        int[] visited = new int[n];
        for (int i = 0; i < n; i++) {
            if (dfs(graph, i, visited)) rets.add(i);
        }
        //Collections.sort(rets);
        return rets;
    }
    private boolean dfs(int[][] graph, int pos, int[] visited) {
        if (visited[pos] == 2) return true;
        if (visited[pos] == 1) return false;

        visited[pos] = 1;
        for (int next : graph[pos]) {
            if (!dfs(graph, next, visited)) return false;
        }
        visited[pos] = 2;
        return true;
    }
}
